pkg/cli/image/manifest/source.go

```diff
@@ -0,0 +1,141 @@
+		icspData, err := ioutil.ReadFile(o.ImageContentSourcePolicyFile)
+		if err != nil {
+			return fmt.Errorf("unable to read ImageContentSourceFile %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		if len(icspData) == 0 {
+			return fmt.Errorf("no data found in ImageContentSourceFile %s", o.ImageContentSourcePolicyFile)
+		}
+		icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+		if err != nil {
+			return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", o.ImageContentSourcePolicyFile, err)
+		}
+		var icsp *operatorv1alpha1.ImageContentSourcePolicy
+		var ok bool
+		if icsp, ok = icspObj.(*operatorv1alpha1.ImageContentSourcePolicy); !ok {
+			return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", o.ImageContentSourcePolicyFile)
+		}
+		o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, *icsp)
+	}
+	o.ICSPClientFn = func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error) {
+		// If ImageContentSourceFile is given, only add ImageContentSource from file, don't search cluster ICSP
+		if len(o.ImageContentSourcePolicyFile) != 0 {
+			return nil, nil
+		}
+		restConfig, err := f.ToRESTConfig()
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		icspClient, err := operatorv1alpha1client.NewForConfig(restConfig)
+		if err != nil {
+			// may or may not be connected to a cluster
+			// don't error if can't connect
+			klog.V(4).Infof("did not connect to an OpenShift 4.x server, will not lookup ImageContentSourcePolicies: %v", err)
+			return nil, nil
+		}
+		return icspClient.ImageContentSourcePolicies(), nil
+	}
+	return nil
+}
+
+// GetICSPs will lookup ICSPs from cluster. Since it's not a hard requirement to find ICSPs from cluster, GetICSPs logs errors rather than returning errors.
+func (o *ImageSourceOptions) GetICSPs(icspClient operatorv1alpha1client.ImageContentSourcePolicyInterface) {
+	icsps, err := icspClient.List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		// may or may not have access to ICSPs in cluster
+		// don't error if can't access ICSPs
+		klog.V(4).Infof("did not access any ImageContentSourcePolicies in cluster: %v", err)
+	}
+	if len(icsps.Items) == 0 {
+		klog.V(4).Info("no ImageContentSourcePolicies found in cluster")
+	}
+	o.ImageContentSourcePolicyList = append(o.ImageContentSourcePolicyList, icsps.Items...)
+}
+
+// GetImageSourcePrefixes returns list of 'registry/repository/name' options to try
+// 'registry/repository/name' determined from examining ImageContentSourcePolicy objects.
+// If a valid prefix is obtained from ICSP, return true
+func GetImageSourcePrefixes(icsps []operatorv1alpha1.ImageContentSourcePolicy, image string) (bool, []string, error) {
```

Comment:
Note that ICSP should work only for images that are referenced by digest. We don't know if the mirrored images are fresh enough and we can trust its tags.

Reply:
Yes, ICSP uses digests only. This change enables extracting from mirrored images and a disconnected registry. The
`registry/repository/name` is substituted.  When mirroring, `registry/repo/image@sha256:123456` becomes `mirroredregistry/repo2/image2@sha256:123456` The sha is the same.  This change enables that substitution during the extract, so you no longer require the auth for the original registry/repo to extract from a mirrored release.

Reply:
Right, but where is the check that the image is referenced by sha?

Reply:
whatever image is user-given (it might be `registry/repo/name:tag` or `registry/repo/name@sha256:digest`) is parsed to use digest, explained here: https://github.com/openshift/oc/blob/d30b2d0a4f1bab7ec8bd51d13c50371c730eee84/pkg/cli/admin/release/info.go#L853-#L896  and verified if the substituted `registry/rep/name@sha256:digest` exists or if it does not exist, the substitution is not made.